pkg/deploy/strategy/rolling/rolling.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/openshift/origin/pkg/deploy/strategy"
+	deployutil "github.com/openshift/origin/pkg/deploy/util"
+)
+
+// TODO: This should perhaps be made public upstream. See:
+// https://github.com/GoogleCloudPlatform/kubernetes/issues/7851
+const sourceIdAnnotation = "kubectl.kubernetes.io/update-source-id"
+
+// RollingDeploymentStrategy is a Strategy which implements rolling
+// deployments using the upstream Kubernetes RollingUpdater.
+//
+// Currently, there are some caveats:
+//
+// 1. When there is no existing prior deployment, deployment delegates to
+// another strategy.
+// 2. The interface to the RollingUpdater is not very clean.
+//
+// These caveats can be resolved with future upstream refactorings to
+// RollingUpdater[1][2].
+//
+// [1] https://github.com/GoogleCloudPlatform/kubernetes/pull/7183
+// [2] https://github.com/GoogleCloudPlatform/kubernetes/issues/7851
+type RollingDeploymentStrategy struct {
+	// initialStrategy is used when there are no prior deployments.
+	initialStrategy strategy.DeploymentStrategy
+	// client is used to deal with ReplicationControllers.
+	client kubectl.RollingUpdaterClient
+	// rollingUpdate knows how to perform a rolling update.
+	rollingUpdate func(config *kubectl.RollingUpdaterConfig) error
+	// codec is used to access the encoded config on a deployment.
+	codec runtime.Codec
+}
+
+// NewRollingDeploymentStrategy makes a new RollingDeploymentStrategy.
+func NewRollingDeploymentStrategy(namespace string, client kclient.Interface, codec runtime.Codec, initialStrategy strategy.DeploymentStrategy) *RollingDeploymentStrategy {
+	updaterClient := &rollingUpdaterClient{
+		GetReplicationControllerFn: func(namespace, name string) (*kapi.ReplicationController, error) {
+			return client.ReplicationControllers(namespace).Get(name)
+		},
+		UpdateReplicationControllerFn: func(namespace string, rc *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+			return client.ReplicationControllers(namespace).Update(rc)
+		},
+		CreateReplicationControllerFn: func(namespace string, rc *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+			return nil, fmt.Errorf("unexpected attempt to create deployment: %#v", rc)
+		},
+		DeleteReplicationControllerFn: func(namespace, name string) error {
+			return fmt.Errorf("unexpected attempt to delete %s/%s", namespace, name)
+		},
+		ControllerHasDesiredReplicasFn: func(rc *kapi.ReplicationController) wait.ConditionFunc {
+			return kclient.ControllerHasDesiredReplicas(client, rc)
+		},
+	}
+	return &RollingDeploymentStrategy{
+		codec:           codec,
+		initialStrategy: initialStrategy,
+		client:          updaterClient,
+		rollingUpdate: func(config *kubectl.RollingUpdaterConfig) error {
+			updater := kubectl.NewRollingUpdater(namespace, updaterClient)
+			return updater.Update(config)
+		},
+	}
+}
+
+func (s *RollingDeploymentStrategy) Deploy(deployment *kapi.ReplicationController, oldDeployments []kapi.ObjectReference) error {
```

Comment:
There was a reason... and I'm struggling to remember it. I'll go ahead and return *ReplicationController refs instead of ObjectReferences.